File: backend/schemas/analytics.py
from datetime import datetime, date
from typing import Optional
from pydantic import BaseModel, ConfigDict


class ListeningHistoryItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    track_id: str
    track_name: str
    artist_id: str
    artist_name: str
    album_id: str
    album_name: str
    album_art_url: Optional[str] = None
    duration_ms: int
    played_at: datetime


class HistoryResponse(BaseModel):
    items: list[ListeningHistoryItem]
    total: int
    page: int
    limit: int


class TodaySummaryResponse(BaseModel):
    total_tracks: int
    listening_time: str
    listening_time_ms: int


class HourBucket(BaseModel):
    hour: int
    count: int


class HourlyVelocityResponse(BaseModel):
    hours: list[HourBucket]


class TopEntity(BaseModel):
    """A top artist or track derived from listening history play counts."""
    id: str
    name: str
    play_count: int
    image_url: Optional[str] = None
    # For tracks only
    artist_name: Optional[str] = None


class OverviewResponse(BaseModel):
    """30-day rollup powering the Dashboard stat row."""
    window_days: int
    total_tracks: int          # total plays in window (rows)
    total_hours: float         # sum(duration_ms) / 3_600_000, rounded to 1 dp
    unique_artists: int
    unique_tracks: int
    top_artist: Optional[TopEntity] = None
    top_track: Optional[TopEntity] = None


class DailyPoint(BaseModel):
    """One day in the trend line."""
    date: date
    plays: int
    minutes: int


class DailyTrendResponse(BaseModel):
    days: int
    points: list[DailyPoint]


class GenreSlice(BaseModel):
    name: str
    weight: int           # 0-100 (percent rounded)
    artist_count: int


class GenreDistributionResponse(BaseModel):
    slices: list[GenreSlice]      # ordered desc by weight; capped to top N + 'Other'
    total_genres: int             # total distinct genres seen across top artists
    based_on_artists: int         # how many top artists were aggregated

File: frontend/src/api/analytics.ts
import api from "@/api/axios";

export const getTopItems = async (
  time_range: string,
  limit: number,
  itemType: string,
) => {
  const response = await api.get(`/analytics/top-items/${itemType}`, {
    params: { time_range, limit },
  });
  return response.data;
};

export const getRecentlyPlayed = async (limit: number) => {
  const response = await api.get("/analytics/recently-played", {
    params: { limit },
  });
  return response.data;
};

export const getUserPlaylists = async (limit: number, offset: number) => {
  const response = await api.get("/analytics/playlists", {
    params: { limit, offset },
  });
  return response.data;
};

export interface TopEntity {
  id: string;
  name: string;
  play_count: number;
  image_url?: string;
  artist_name?: string;
}

export interface OverviewResponse {
  window_days: number;
  total_tracks: number;
  total_hours: number;
  unique_artists: number;
  unique_tracks: number;
  top_artist: TopEntity | null;
  top_track: TopEntity | null;
}

export interface DailyPoint {
  date: string; // ISO date "YYYY-MM-DD"
  plays: number;
  minutes: number;
}

export interface DailyTrendResponse {
  days: number;
  points: DailyPoint[];
}

export interface GenreSlice {
  name: string;
  weight: number;
  artist_count: number;
}

export interface GenreDistributionResponse {
  slices: GenreSlice[];
  total_genres: number;
  based_on_artists: number;
}

export interface TodaySummaryResponse {
  total_tracks: number;
  listening_time: string;
  listening_time_ms: number;
}

export interface HourBucket {
  hour: number;
  count: number;
}

export interface HourlyVelocityResponse {
  hours: HourBucket[];
}

export interface HistoryItem {
  id: number;
  track_id: string;
  track_name: string;
  artist_id: string;
  artist_name: string;
  album_id: string;
  album_name: string;
  album_art_url?: string;
  duration_ms: number;
  played_at: string; // ISO datetime
}

export const getOverview = async (days = 30): Promise<OverviewResponse> => {
  const response = await api.get("/deep-analytics/overview", { params: { days } });
  return response.data;
};

export const getDailyTrend = async (days = 30): Promise<DailyTrendResponse> => {
  const response = await api.get("/deep-analytics/daily-trend", {
    params: { days },
  });
  return response.data;
};

export const getGenreDistribution = async (
  top_n = 5,
): Promise<GenreDistributionResponse> => {
  const response = await api.get("/deep-analytics/genre-distribution", {
    params: { top_n },
  });
  return response.data;
};

export const getTodaySummary = async (): Promise<TodaySummaryResponse> => {
  const response = await api.get("/deep-analytics/summary/today");
  return response.data;
};

export const getHourlyVelocity = async (): Promise<HourlyVelocityResponse> => {
  const response = await api.get("/deep-analytics/hourly-velocity");
  return response.data;
};

export interface HistoryResponse {
  items: HistoryItem[];
  total: number;
  page: number;
  limit: number;
}

export const getHistory = async (
  page: number,
  limit: number,
): Promise<HistoryResponse> => {
  const response = await api.get("/deep-analytics/history", {
    params: { page, limit },
  });
  return response.data;
};
